### PR TITLE
Add an Hook that converts user defined types to JDBC primitive types

### PR DESCRIPTION
**Pull request by gakuzzzz**
## Motivation

Currently, ScalikeJDBC does not support the conversion user defined types to JDBC primitive types.

So, it can compile following code, but it does not work as expected.

```scala
case class UserName(value: String)

...

def insertUser(name: UserName)(implicit session: DBSession): User = {
  applyUpdate {
    insertInto(User).namedValues(
      u.name -> name // this is mistake. currently it should be `name.value`
    )
  }

}
```

this PR introduce `ParameterBinderFactory` type class.

It converts user defined types.

## Notation

this PR breaks the binary compatibility.

TODO

* [x] tests
* [x] Scala2.10.x support
* [x] rebase & squash

### CODE REVIEW

File: scalikejdbc-core/src/main/scala/scalikejdbc/TypeUnbinder.scala [hunk 1]
```diff
@@ -0,0 +1,113 @@
+package scalikejdbc
+
+import java.io.InputStream
+import java.sql.PreparedStatement
+import scalikejdbc.UnixTimeInMillisConverterImplicits._
+import scalikejdbc.interpolation.SQLSyntax
+import scala.language.experimental.macros
+import scala.reflect.macros.blackbox.Context
+
+case class UnbindResult(value: Either[SQLSyntax, ParameterBinder])
```
Comment (seratch) [inline scalikejdbc-core/src/main/scala/scalikejdbc/TypeUnbinder.scala]: UnboundResult?

Comment (gakuzzzz) [inline scalikejdbc-core/src/main/scala/scalikejdbc/TypeUnbinder.scala]: ya!

File: project/Build.scala [hunk 2]
```diff
@@ -37,6 +37,7 @@ object ScalikeJDBCProjects extends Build {
     publishTo <<= version { (v: String) => _publishTo(v) },
     publishMavenStyle := true,
     resolvers ++= _resolvers,
+    resolvers in Global += "pr" at "https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/",
```
Comment (xuwei-k) [inline project/Build.scala]: unnecessary change

File: scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+private[scalikejdbc] case class SQLSyntaxParameterBinder(syntax: SQLSyntax) extends ParameterBinder[SQLSyntax] {
+  val value = syntax
+  def apply(stmt: PreparedStatement, idx: Int): Unit = ()
+}
+
+trait ParameterBinderFactory[A] { self =>
+
+  def apply(value: A): ParameterBinder[A]
+
+  def bimap[B](f: A => B, g: B => A): ParameterBinderFactory[B] = new ParameterBinderFactory[B] {
```
Comment (xuwei-k) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: `xmap` or `inmap` ?

- https://github.com/scalaz/scalaz/blob/v7.2.1/core/src/main/scala/scalaz/InvariantFunctor.scala#L27
- https://github.com/playframework/playframework/blob/2.5.0/framework/src/play-functional/src/main/scala/play/api/libs/functional/Functors.scala#L26

Comment (gakuzzzz) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: you're right!

in cats, `imap` https://github.com/typelevel/cats/blob/b638e298ca544739602e76388c9d266c4bc4357d/core/src/main/scala/cats/functor/Invariant.scala#L10

Should we choose them?

Comment (gakuzzzz) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: I change it to `xmap`

File: scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala [hunk 4]
```diff
@@ -475,11 +475,12 @@ trait QueryDSLFeature { self: SQLInterpolationFeature with SQLSyntaxSupportFeatu
   case class InsertSQLBuilder(override val sql: SQLSyntax) extends SQLBuilder[UpdateOperation] {
 
     def columns(columns: SQLSyntax*): InsertSQLBuilder = this.copy(sql = sqls"${sql} (${sqls.csv(columns: _*)})")
+    @deprecated(message = "values is not type safe. use `namedValues` instead.", since = "2.4.0")
```
Comment (seratch) [inline scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala]: I don't prefer this deprecation. The `namedValues` with `EssentialParameterBinder` is not fully type safe in practice. `ParameterBinder` doesn't guarantee parameter's type to be compatible with its column type.

File: scalikejdbc-interpolation/src/test/scala/scalikejdbc/QueryInterfaceSpec.scala [hunk 5]
```diff
@@ -353,7 +353,7 @@ class QueryInterfaceSpec extends FlatSpec with Matchers with DBSettings with SQL
         {
           val notInClauseResults = withSQL {
             select.from(Order as o)
-              .where.not.in(o.id, Seq())
+              .where.not.in(o.id, Seq[Int]())
```
Comment (seratch) [inline scalikejdbc-interpolation/src/test/scala/scalikejdbc/QueryInterfaceSpec.scala]: In general, it's ok for us, well-trained Scala programmers, but OTOH, it may be a bit confusing error for Scala beginners. Is there any way to improve the error message to suggest what to do?

```scala
[error] src/test/scala/scalikejdbc/QueryInterfaceSpec.scala:333: ambiguous implicit values:
[error]  both value sqlArrayParameterBinderFactory in object ParameterBinderFactory of type => scalikejdbc.ParameterBinderFactory[java.sql.Array]
[error]  and value bigDecimalParameterBinderFactory in object ParameterBinderFactory of type => scalikejdbc.ParameterBinderFactory[BigDecimal]
[error]  match expected type scalikejdbc.ParameterBinderFactory[B]
[error]               .where.in(o.id, Seq.empty)
[error]                        ^
```

File: scalikejdbc-core/src/test/scala/scalikejdbc/interpolation/SQLSyntaxSpec.scala [hunk 6]
```diff
@@ -16,7 +22,7 @@ class SQLSyntaxSpec extends FlatSpec with Matchers {
   it should "have #append" in {
     val s = SQLSyntax.eq(sqls"id", 123).append(sqls"and name is not null")
     s.value should equal(" id = ? and name is not null")
-    s.parameters should equal(Seq(123))
+    s.parameters.map(parameterExtractor) should equal(Seq(123))
```
Comment (seratch) [inline scalikejdbc-core/src/test/scala/scalikejdbc/interpolation/SQLSyntaxSpec.scala]: hmm, this is not a bin-incompatibility but an API breaking change. I'm not sure this is safe enough yet.

File: scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinder.scala [hunk 7]
```diff
@@ -15,19 +15,26 @@ import java.sql.PreparedStatement
  * sql"insert into table (bin) values (${bin})".update.apply()
  * }}}
  */
-trait ParameterBinder[A] {
+trait EssentialParameterBinder { self =>
```
Comment (seratch) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinder.scala]: Can we hide this trait from library users? `Essential` looks a bit Play Framework-ish for me. Any other good name for this? Just `Parameter`?

File: scalikejdbc-interpolation/src/test/scala/scalikejdbc/QueryInterfaceSpec.scala (hunk 5 above)
Comment (xuwei-k) [inline scalikejdbc-interpolation/src/test/scala/scalikejdbc/QueryInterfaceSpec.scala]: Maybe we can use same trick in `TxBoundaryMissingImplicits` ?

https://github.com/scalikejdbc/scalikejdbc/blob/2.3.5/scalikejdbc-core/src/main/scala/scalikejdbc/TxBoundary.scala#L46-L47

File: scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  implicit def optionalParameterBinderFactory[A](implicit ev: ParameterBinderFactory[A]): ParameterBinderFactory[Option[A]] = new ParameterBinderFactory[Option[A]] {
+    def apply(value: Option[A]): ParameterBinder[Option[A]] = {
+      if (value == null) ParameterBinder.NullParameterBinder
+      else value.fold[ParameterBinder[Option[A]]](ParameterBinder.NullParameterBinder)(v => ev(v).map(Option.apply))
+    }
+  }
+
+  def jsr310ParameterBinderFactory[A]: ParameterBinderFactory[A] = ParameterBinderFactory[A] { p => (underlying, i) =>
+    // Accessing JSR-310 APIs via Java reflection
+    // because scalikejdbc-core should work on not only Java 8 but 6 & 7.
```
Comment (xuwei-k) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: I think we can avoid reflections if define `jsr310ParameterBinderFactory` in `scalikejdbc-jsr310` module?

Comment (gakuzzzz) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: I agree. I think it should move scalikejdbc-jsr310.

However, If it do so, it breaks compatibility heavily. https://github.com/scalikejdbc/scalikejdbc/blob/82b3541529439b906d709b64164dba1c90f17277/scalikejdbc-core/src/main/scala/scalikejdbc/StatementExecutor.scala#L92-L114

Does an migration guide resolve this?

File: scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala (hunk 4 above)
Comment (gakuzzzz) [inline scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala]: I agree that it is not fully type safe in practice.

However, I think that this method should be deprecated.
Or change the behavior as throwing an exception instead of using `setObject`.

Current behavior causes many troubles.

Comment (seratch) [inline scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala]: I can understand your opinion but I have no plan to remove this method. So I cannot agree to mark this method as deprecated.

File: scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala (hunk 8 above)
Comment (seratch) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: You must keep StatementExecutor as-is in scalikejdbc 2.4. Is it impossible to move this `jsr310ParameterBinderFactory` to scalikejdbc-jsr310 library? IMO, it's also fine to have this here.

These implementation that use Java reflection will be replaced since scalikejdb libs 3.0. scalikejdbc 3.0 will ...

- support scala 2.12 too
- drop java7 support (scala 2.12 doesn't support java7)
- merge scalikejdbc-jsr310 into scalikejdbc-core

**Comment by seratch** [comment]
Quite nice improvement. Please check my comments, fix them and squash commits. After that, I'm willing to merge this and am going to prepare for scalikejdbc 2.4 series release.

File: scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala (hunk 8 above)
Comment (gakuzzzz) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: The other hand, If we write `ParameterBinderFactory[java.time.XXX]`s at inside ParameterBinderFactoryMacro.any, we can avoid runtime reflections.

Comment (xuwei-k) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: > However, If it do so, it breaks compatibility heavily.

I **don't** thinks so.

Comment (cppexpert) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: Frankly I think whole `ParameterBinder` is extremely inconvenient to use and verbose. I doubt automagic hook will be any better. I ended up creating extension object with implicit converters that looks like `.toSqlType` method for anything. I can open source it if you guys think it's a good idea. I've also created extensions to support Postgress jsonb and array functions/operators in ScalikeJDBC DSL.

Comment (seratch) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinderFactory.scala]: It is very great if you share your work! But this pull request should be also a fine improvement. That's another story.

File: scalikejdbc-core/src/main/scala/scalikejdbc/SQL.scala [hunk 9]
```diff
@@ -5,7 +5,7 @@ import scala.language.higherKinds
 import scala.collection.generic.CanBuildFrom
 
 /**
- * SQL abstraction's companion object.
+ * SQL abstraction's companirawParameters
```
Comment (xuwei-k) [inline scalikejdbc-core/src/main/scala/scalikejdbc/SQL.scala]: ?

Comment (gakuzzzz) [inline scalikejdbc-core/src/main/scala/scalikejdbc/SQL.scala]: What is this !?!?!?
maybe mistake of replace. I'll fix it.

File: scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala (hunk 4 above)
Comment (gakuzzzz) [inline scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala]: Are there other good way how protect the type conversion fail instead of deprecated annotation ?

Does it throw an exception at StatementExecutor?

File: scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinder.scala [hunk 10]
```diff
@@ -15,20 +15,27 @@ import java.sql.PreparedStatement
...
    * Applies parameter to PreparedStatement.
    */
   def apply(stmt: PreparedStatement, idx: Int): Unit
 
 }
+trait ParameterBinderWithValue[A] extends ParameterBinder { self =>
+
+  def value: A
+
+  // [[scalikejdbc.ParameterBinder#map]] breaks the Functor-law
```
Comment (xuwei-k) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinder.scala]: Is this the reason of `private[scalikejdbc]` ?

File: scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala (hunk 4 above)
Comment (seratch) [inline scalikejdbc-interpolation/src/main/scala/scalikejdbc/QueryDSLFeature.scala]: If we can have a complete white list of types to accept for all existing JDBC drivers, throwing exception when surely invalid pameters are passed is fine. Is it really possible?

File: scalikejdbc-interpolation/src/test/scala/scalikejdbc/QueryInterfaceSpec.scala (hunk 5 above)
Comment (seratch) [inline scalikejdbc-interpolation/src/test/scala/scalikejdbc/QueryInterfaceSpec.scala]: My suggestion here is just a nice-to-have thing. Ignore it if there is no solution for now.

File: scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinder.scala (hunk 10 above)
Comment (gakuzzzz) [inline scalikejdbc-core/src/main/scala/scalikejdbc/ParameterBinder.scala]: this method is dirty hack for logging/ParameterBinderFactory#xmap.
And it breaks the Functor-law.

so I dont think that it should be public to ScalikeJDBC users.

**Comment by seratch** [comment]
Thanks. LGTM 👍
